crates/matrix-sdk-ui/src/timeline/builder.rs

```diff
@@ -205,11 +206,20 @@ impl TimelineBuilder {
             // The events we're injecting might be encrypted events, but we might
             // have received the room key to decrypt them while nobody was listening to the
             // `m.room_key` event, let's retry now.
-            //
-            // TODO: We could spawn a task here and put this into the background, though it
-            // might not be worth it depending on the number of events we injected.
-            // Some measuring needs to be done.
-            timeline.retry_decryption_for_all_events().await;
+
+            // If there is more than 5 events to decrypt, let's spawn a task so that it's
+            // not blocking the main flow too much. Everything is async here, but if someone
+            // is calling the `Timeline` builder in a blocking context, it could block for
+            // too long.
+            if number_of_events > 5 {
```

Comment:
Instead of checking the number of events, can we check whether there are any in UTD state? (I think we should spawn for 1-4 as well, not just 5+, spawning a task should be faster than retrying even a single decryption)